// @redwoodjs/cli/dist/commands/setup/cache/cache.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = exports.description = exports.command = exports.builder = void 0;
var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/promise"));
var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/interopRequireWildcard"));
var _terminalLink = _interopRequireDefault(require("terminal-link"));
const command = 'cache <client>';
exports.command = command;
const description = 'Sets up an init file for service caching';
exports.description = description;
const builder = yargs => {
  yargs.positional('client', {
    choices: ['memcached', 'redis'],
    description: 'Cache client',
    type: 'string',
    required: true
  }).option('force', {
    alias: 'f',
    default: false,
    description: 'Overwrite existing cache.js file',
    type: 'boolean'
  }).epilogue(`Also see the ${(0, _terminalLink.default)('Redwood CLI Reference', 'https://redwoodjs.com/docs/cli-commands#setup-cache')}`);
};
exports.builder = builder;
const handler = async options => {
  const {
    handler
  } = await _promise.default.resolve().then(() => (0, _interopRequireWildcard2.default)(require('./cacheHandler')));
  return handler(options);
};
exports.handler = handler;
