// @redwoodjs/graphql-server/dist/errors.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidationError = exports.UserInputError = exports.SyntaxError = exports.RedwoodGraphQLError = exports.PersistedQueryNotSupportedError = exports.PersistedQueryNotFoundError = exports.ForbiddenError = exports.AuthenticationError = void 0;
var _setPrototypeOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/object/set-prototype-of"));
var _graphql = require("graphql");
// based on ApolloError https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-errors/src/index.ts

class RedwoodGraphQLError extends _graphql.GraphQLError {
  constructor(message, extensions, originalError) {
    super(message, {
      extensions: {
        ...extensions,
        code: extensions?.code || 'REDWOODJS_ERROR'
      },
      originalError
    });
    (0, _setPrototypeOf.default)(this, _graphql.GraphQLError.prototype);
  }
}
exports.RedwoodGraphQLError = RedwoodGraphQLError;
class SyntaxError extends RedwoodGraphQLError {
  constructor(message) {
    super(message, {
      code: 'GRAPHQL_PARSE_FAILED'
    });
    (0, _setPrototypeOf.default)(this, SyntaxError.prototype);
  }
}
exports.SyntaxError = SyntaxError;
class ValidationError extends RedwoodGraphQLError {
  constructor(message) {
    super(message, {
      code: 'GRAPHQL_VALIDATION_FAILED'
    });
    (0, _setPrototypeOf.default)(this, ValidationError.prototype);
  }
}
exports.ValidationError = ValidationError;
class AuthenticationError extends RedwoodGraphQLError {
  constructor(message) {
    super(message, {
      code: 'UNAUTHENTICATED'
    });
    (0, _setPrototypeOf.default)(this, AuthenticationError.prototype);
  }
}
exports.AuthenticationError = AuthenticationError;
class ForbiddenError extends RedwoodGraphQLError {
  constructor(message) {
    super(message, {
      code: 'FORBIDDEN'
    });
    (0, _setPrototypeOf.default)(this, ForbiddenError.prototype);
  }
}
exports.ForbiddenError = ForbiddenError;
class PersistedQueryNotFoundError extends RedwoodGraphQLError {
  constructor() {
    super('PersistedQueryNotFound', {
      code: 'PERSISTED_QUERY_NOT_FOUND'
    });
    (0, _setPrototypeOf.default)(this, PersistedQueryNotFoundError.prototype);
  }
}
exports.PersistedQueryNotFoundError = PersistedQueryNotFoundError;
class PersistedQueryNotSupportedError extends RedwoodGraphQLError {
  constructor() {
    super('PersistedQueryNotSupported', {
      code: 'PERSISTED_QUERY_NOT_SUPPORTED'
    });
    (0, _setPrototypeOf.default)(this, PersistedQueryNotSupportedError.prototype);
  }
}
exports.PersistedQueryNotSupportedError = PersistedQueryNotSupportedError;
class UserInputError extends RedwoodGraphQLError {
  constructor(message, properties) {
    super(message, {
      code: 'BAD_USER_INPUT',
      properties
    });
    (0, _setPrototypeOf.default)(this, UserInputError.prototype);
  }
}
exports.UserInputError = UserInputError;
