// @reach/skip-nav/dist/reach-skip-nav.cjs.dev.js
"use strict";
/**
  * @reach/skip-nav v0.18.0
  *
  * Copyright (c) 2018-2022, React Training LLC
  *
  * This source code is licensed under the MIT license found in the
  * LICENSE.md file in the root directory of this source tree.
  *
  * @license MIT
  */

var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/reach-skip-nav.tsx
var reach_skip_nav_exports = {};
__export(reach_skip_nav_exports, {
  SkipNavContent: () => SkipNavContent,
  SkipNavLink: () => SkipNavLink
});
module.exports = __toCommonJS(reach_skip_nav_exports);
var React = __toESM(require("react"));
var defaultId = "reach-skip-nav";
var SkipNavLink = React.forwardRef(function SkipNavLink2(_a, forwardedRef) {
  var _b = _a, { as: Comp = "a", children = "Skip to content", contentId } = _b, props = __objRest(_b, ["as", "children", "contentId"]);
  let id = contentId || defaultId;
  return /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({}, props), {
    ref: forwardedRef,
    href: `#${id}`,
    "data-reach-skip-link": "",
    "data-reach-skip-nav-link": ""
  }), children);
});
SkipNavLink.displayName = "SkipNavLink";
var SkipNavContent = React.forwardRef(function SkipNavContent2(_a, forwardedRef) {
  var _b = _a, { as: Comp = "div", id: idProp } = _b, props = __objRest(_b, ["as", "id"]);
  let id = idProp || defaultId;
  return /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({}, props), {
    ref: forwardedRef,
    id,
    "data-reach-skip-nav-content": ""
  }));
});
SkipNavContent.displayName = "SkipNavContent";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SkipNavContent,
  SkipNavLink
});
