// @redwoodjs/graphql-server/dist/plugins/useArmor.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.useArmor = void 0;
var _graphqlArmor = require("@escape.tech/graphql-armor");
const armorConfigDefaultOptions = {
  logContext: false,
  logErrors: true
};
const useArmor = (logger, config) => {
  const logRejection = (ctx, error) => {
    if (config?.logContext ?? armorConfigDefaultOptions.logContext) {
      if (ctx) {
        logger.debug({
          custom: ctx
        }, `Armor rejection context`);
      }
    }
    if (config?.logErrors ?? armorConfigDefaultOptions.logErrors) {
      logger.error(error, `Armor rejected request: ${error.message}`);
    }
  };
  return (0, _graphqlArmor.EnvelopArmorPlugin)({
    ...config,
    costLimit: {
      ...config?.costLimit,
      onReject: [logRejection]
    },
    maxAliases: {
      ...config?.maxAliases,
      onReject: [logRejection]
    },
    maxDepth: {
      ...config?.maxDepth,
      onReject: [logRejection]
    },
    maxDirectives: {
      ...config?.maxDirectives,
      onReject: [logRejection]
    },
    maxTokens: {
      ...config?.maxTokens,
      onReject: [logRejection]
    }
  });
};
exports.useArmor = useArmor;
